File: csp/23/4007.非零段划分/4007.非零段划分.cpp
// 暴力

// #include <iostream>
// #include <cstring>
// #include <cstdio>
// #include <algorithm>

// using namespace std;

// const int N = 5 * 1e5 + 10;

// int a[N];

// int main()
// {
//     int n;
//     cin >> n;
//     for (int i = 1; i <= n; i++)
//         cin >> a[i];

//     int res = 0;
//     for (int p = 1; p <= 1e4; p++)
//     {
//         int cnt = 0;
//         for (int i = 1; i <= n; i++)
//         {
//             if (a[i] < p)
//                 continue;
//             int j = i;
//             cnt += 1;
//             while (j <= n && a[j + 1] >= p)
//             {
//                 j++;
//             }
//             i = j + 1;
//         }
//         res = max(res, cnt);
//     }
//     cout << res << endl;
//     return 0;
// }

#include <iostream>
#include <cstring>
#include <cstdio>
#include <algorithm>

using namespace std;

const int N = 5 * 1e5 + 10, M = 1e4 + 10;

int a[N], cnt[N];

int main()
{
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> a[i];
    n = unique(a + 1, a + 1 + n) - a - 1;
    a[0] = 0, a[n + 1] = 0;
    for (int i = 1; i <= n; i++)
    {
        int x = a[i - 1], y = a[i], z = a[i + 1];
        if (x < y && y > z)
            cnt[y] += 1;
        else if (x > y && y < z)
            cnt[y] -= 1;
    }
    int res = 0;
    int sum = 0;
    for (int i = M; i; i--)
    {
        sum += cnt[i];
        res = max(res, sum);
    }
    cout << res << endl;
    return 0;
}

File: csp/23/4007.非零段划分/4007.非零段划分.py
N = 5 * int(1e5) + 10
M = int(1e5) + 10

cnt = [0] * N

n = int(input().strip())
a = list(map(int, input().strip().split()))
a = [v for i, v in enumerate(a) if i == 0 or v != a[i - 1]]

n = len(a)
a = [0] + a + [0]

for i in range(1, n + 1):
    x = a[i - 1]
    y = a[i]
    z = a[i + 1]
    if x > y and y < z:
        cnt[y] -= 1
    elif x < y and y > z:
        cnt[y] += 1

res = 0
SUM = 0
for i in range(M, 0, -1):
    SUM += cnt[i]
    res = max(res, SUM)

print(res)
